HopsanGUI/Widgets/ModelWidget.cpp

```diff
@@ -698,16 +698,260 @@ void ModelWidget::saveAs()
+                        UnitConverter uc2 = gpConfig->getUnitScaleUC(hopsanQuantity, parameter.mUnit); //Converts between Hopsan unit and base unit
+                        pSystem->setParameterValue(parameterName, uc2.convertFromBase(uc1.convertToBase(ssvValue)));
+                    }
+                }
+            }
+            else {
+                pSystem->setParameterValue(parameterName, ssvValue);   //Ignore units for non-Real parameters
+            }
+        }
+        else {                               //Set component parameter
+            ModelObject *pComponent = pSystem->getModelObject(componentName);
+            if(pComponent == nullptr) {
+                gpMessageHandler->addWarningMessage("Could not find component: "+componentName+", ignoring parameter "+parameterName);
+                parameterElement = parameterElement.nextSiblingElement("ssv:Parameter");
+                continue;
+            }
+            CoreParameterData parameter;
+            pComponent->getParameter(parameterName, parameter);
+            if(dataTypeTranslator[parameter.mType] != ssvDataType) {
+                gpMessageHandler->addWarningMessage("Wrong data type for unit: \""+ssvName+"\", ignoring value.");
+            }
+            else if(dataTypeTranslator[parameter.mType] == "Real") {
+                if(parameter.mUnit == ssvUnit) {
+                    pComponent->setParameterValue(parameterName, ssvValue);
+                }
+                else {
+                    QStringList quantities = gpConfig->getQuantitiesForUnit(ssvUnit);
+                    if(quantities.isEmpty()) {
+                        gpMessageHandler->addWarningMessage("Unit is not supported by Hopsan: "+ssvUnit+". Setting parameter without unit conversion.");
+                        pComponent->setParameterValue(parameterName, ssvValue);
+                    }
+                    else {
+                        QString ssvQuantity, hopsanUnit, hopsanQuantity;
+                        ssvQuantity = quantities[0];    //Guessing for first quantity is the best we can do
+                        hopsanUnit = parameter.mUnit;
+                        if(parameter.mQuantity.isEmpty()) {
+                            hopsanQuantity = gpConfig->getQuantitiesForUnit(parameter.mUnit)[0];
+                        }
+                        else {
+                            hopsanQuantity = parameter.mQuantity;
+                        }
+                        UnitConverter uc1 = gpConfig->getUnitScaleUC(ssvQuantity, ssvUnit);         //Converts between SSV unit and base unit
+                        UnitConverter uc2 = gpConfig->getUnitScaleUC(hopsanQuantity, parameter.mUnit); //Converts between Hopsan unit and base unit
+                        pComponent->setParameterValue(parameterName, uc2.convertFromBase(uc1.convertToBase(ssvValue)));
```

Comment:
I am unsure about this, but should we really convert from base, should we not always set parameters in the base unit.
Or is the conversion handled inside pComponent->setParameterValue? In this case I believe that pComponent is a GUIComponent. So maybe that is the case.

In the core I think the idea is that we always set parameter values in base unit. But I do not remember really.

Reply:
We need to set variable in the unit specified in the GUIComponent. This value is then converted again in the core. There is a slight risk that we lose precision, but I guess that is unavoidable.

Reply:
I think I know why I still feel this is wrong (it probably is correct).
As I remember it, in the core (component c++ implementation)  units "must" be in base units. 

When you add a parameter or in/out variable you either give a unit or a quantity. 
The old way is to give it a unit, but then you can not be sure how to convert it unless there is a unique match among one of the quantities. That is why we have that odd code allover the place that tries to reverse lookup the quantity.

The "new" way was to give it a quantity, then the value is assumed to be in base unit. Because of interoperability with other components. If you add an output variable with quantity "position" than any receiver of that variable must expect "position" and use the agreed upon unit (in this case base-unit) because we use precompiled components so we can not adapt if someone uses for instance mm instead of m.

So in this case if a component is made correctly, the core unit "must" be the base unit of its quantity (if a quantity was specified instead of unit). Therefor the uc2 converter looked wrong to me, because when you set the parameter value here it goes directly into core without conversion. (Core does not know about unit converters (but it would be nice if it did)).

But in this case, the uc2 converter will most often be 1:1 so it does not matter.
And in the case that someone has actually chosen a different unit (using the old method of directly specifying unit), for instance "mm" for a position variable then this uc2 is actually needed.

Does my reasoning make sense?

Extra:
One can also argue whether choosing a variable (start-value) or constant parameter to have a non base unit, is a good or bad idea. For constant parameters it make more sense since they are never connected to anything else.

I think we came to the conclusion a long time ago that it would be nice if you could in the core choose quantity and unit.
At least for constant parameters, like "diameter = Length:mm" to specify both quantity and unit, but then core must handle all conversion (which would be nice) and it would be possible to discover if someone connects "mm" to "m" for instance. But it would be a headache for users if different precompiled components use different units in variable data (signal/power connectors) . 
That is why I think my idea of requiring base units in the core makes sense. You only do the conversion in the GUI when you set variables or analyze results, as a user convenience.

Reply:
I believe reasoning is correct. There could be cases where you really want to have a non-SI unit as a value, for example both Hz and rad/s are valid units for frequency. There are also certain domains, e.g. aerospace engineering, where different units are sometimes used as a way to avoid mixing up different parameters. In such cases the solution with both quantity and unit would make sense. A problem is that units are stored in the configuration file at the moment, which is not read by the core.